### PR TITLE
Fix/soundwire realtek codec inits

### PR DESCRIPTION
**Pull request by plbossart**
We need to avoid enabling pm_runtime multiple times.

Open: do we actually need both hw_init and first_init, or can we do all the codec initializations only once. When the enumeration restart the codec has typically lost all context.

@shumingfan @bardliao , please comment

### CODE REVIEW

**Comment by bardliao** [comment]
@shumingfan Could you clarify when should we do io_init? Will codec registers be reset to default when it is unattached?

**Comment by shumingfan** [comment]
> arify when should we do

@bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once. 
When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
But, the codec could keep the settings in which set by io_init function.
I could double confirm whether the codec loses any settings of io_init after the enumeration.

File: sound/soc/codecs/rt700.c [hunk 1]
```diff
@@ -1240,10 +1241,15 @@ int rt700_io_init(struct device *dev, struct sdw_slave *slave)
...
-	pm_runtime_set_autosuspend_delay(&slave->dev, 3000);
-	pm_runtime_use_autosuspend(&slave->dev);
-	pm_runtime_mark_last_busy(&slave->dev);
-	pm_runtime_enable(&slave->dev);
+	if (!rt700->first_init) {
+		pm_runtime_set_autosuspend_delay(&slave->dev, 3000);
+		pm_runtime_use_autosuspend(&slave->dev);
+		pm_runtime_mark_last_busy(&slave->dev);
+		pm_runtime_enable(&slave->dev);
+		rt700->first_init = true;
```
Comment (RanderWang) [inline sound/soc/codecs/rt700.c]: one question: can we use "bool pm_runtime_enabled(struct device *dev)" instead of first_init ?

Comment (plbossart) [reply sound/soc/codecs/rt700.c]: @ranj063 can you help here? It's my understanding that the user can change the values and enable/disable pm_runtime via debugfs, so using the pm_runtime status to check if we need to enable pm_runtime seems a bit hairy.

**Comment by shumingfan** [comment]
> > arify when should we do
> 
> @bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once.
> When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
> But, the codec could keep the settings in which set by io_init function.
> I could double confirm whether the codec loses any settings of io_init after the enumeration.

@bardliao @plbossart  I had checked the context of the registers set by io_init (rt700/rt711/rt1308) after the enumeration.
As I said, the codec will keep the context of implementation defined registers.
I think the driver could do io_init only once when codec initialization.

**Comment by plbossart** [comment]
> > > arify when should we do
> > 
> > 
> > @bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once.
> > When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
> > But, the codec could keep the settings in which set by io_init function.
> > I could double confirm whether the codec loses any settings of io_init after the enumeration.
> 
> @bardliao @plbossart I had checked the context of the registers set by io_init (rt700/rt711/rt1308) after the enumeration.
> As I said, the codec will keep the context of implementation defined registers.
> I think the driver could do io_init only once when codec initialization.

Are you sure? before the enumeration starts, the master will issue a Severe Reset (Hard Reset before v1.2) with the alternating pattern of 4096+ 0/1s, and that will force the device to reset. Are you saying the imp-def registers are not affected by a Severe Reset?

**Comment by shumingfan** [comment]
> > > > arify when should we do
> > > 
> > > 
> > > @bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once.
> > > When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
> > > But, the codec could keep the settings in which set by io_init function.
> > > I could double confirm whether the codec loses any settings of io_init after the enumeration.
> > 
> > 
> > @bardliao @plbossart I had checked the context of the registers set by io_init (rt700/rt711/rt1308) after the enumeration.
> > As I said, the codec will keep the context of implementation defined registers.
> > I think the driver could do io_init only once when codec initialization.
> 
> Are you sure? before the enumeration starts, the master will issue a Severe Reset (Hard Reset before v1.2) with the alternating pattern of 4096+ 0/1s, and that will force the device to reset. Are you saying the imp-def registers are not affected by a Severe Reset?

@plbossart Yes, in use case currently, the clockstopmode1 reset just affects MIPI-defined registers.
The imp-def registers will keep the context except for power-on-reset or internal reset command.
Do you have any playback/recording problems after the enumeration without more io_init calls?

**Comment by plbossart** [comment]
> > > > > arify when should we do
> > > > 
> > > > 
> > > > @bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once.
> > > > When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
> > > > But, the codec could keep the settings in which set by io_init function.
> > > > I could double confirm whether the codec loses any settings of io_init after the enumeration.
> > > 
> > > 
> > > @bardliao @plbossart I had checked the context of the registers set by io_init (rt700/rt711/rt1308) after the enumeration.
> > > As I said, the codec will keep the context of implementation defined registers.
> > > I think the driver could do io_init only once when codec initialization.
> > 
> > 
> > Are you sure? before the enumeration starts, the master will issue a Severe Reset (Hard Reset before v1.2) with the alternating pattern of 4096+ 0/1s, and that will force the device to reset. Are you saying the imp-def registers are not affected by a Severe Reset?
> 
> @plbossart Yes, in use case currently, the clockstopmode1 reset just affects MIPI-defined registers.
> The imp-def registers will keep the context except for power-on-reset or internal reset command.
> Do you have any playback/recording problems after the enumeration without more io_init calls?

I was not talking about the clockstopmode1 reset, but the "Bus Reset" (Table 39 of SoundWire spec 1.1, Table 44 of spec 1.2). On S3 resume, the Master starts with a Bus Reset and that *should* have the effect of resetting everything. This Bus Reset helps make sure the lines are in the right electrical level and that the codec is in a known state. When we implement clock-stop, we will use it only for pm_runtime. We have no plans to support clockstopmode1 for now.

**Comment by shumingfan** [comment]
> > > > > > arify when should we do
> > > > > 
> > > > > 
> > > > > @bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once.
> > > > > When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
> > > > > But, the codec could keep the settings in which set by io_init function.
> > > > > I could double confirm whether the codec loses any settings of io_init after the enumeration.
> > > > 
> > > > 
> > > > @bardliao @plbossart I had checked the context of the registers set by io_init (rt700/rt711/rt1308) after the enumeration.
> > > > As I said, the codec will keep the context of implementation defined registers.
> > > > I think the driver could do io_init only once when codec initialization.
> > > 
> > > 
> > > Are you sure? before the enumeration starts, the master will issue a Severe Reset (Hard Reset before v1.2) with the alternating pattern of 4096+ 0/1s, and that will force the device to reset. Are you saying the imp-def registers are not affected by a Severe Reset?
> > 
> > 
> > @plbossart Yes, in use case currently, the clockstopmode1 reset just affects MIPI-defined registers.
> > The imp-def registers will keep the context except for power-on-reset or internal reset command.
> > Do you have any playback/recording problems after the enumeration without more io_init calls?
> 
> I was not talking about the clockstopmode1 reset, but the "Bus Reset" (Table 39 of SoundWire spec 1.1, Table 44 of spec 1.2). On S3 resume, the Master starts with a Bus Reset and that _should_ have the effect of resetting everything. This Bus Reset helps make sure the lines are in the right electrical level and that the codec is in a known state. When we implement clock-stop, we will use it only for pm_runtime. We have no plans to support clockstopmode1 for now.

I checked with HW engineer again. The rt700/rt711/rt1308 follow SdW spec. 1.1. So I focus on SdW spec. 1.1.
According to Table 39, BusReset belongs to the Hard of severity. It is the same with clockstopmode1 reset. The HW follows the Table 40 to make the effect of reset in slave. These effects don't include imp-def registers. Hence, I tested again with BusReset, the imp-def still keeps the context.
If there is a sentence to indicate the imp-def needs to reset, please tell me. I can notice the HW engineer to check.

**Comment by plbossart** [comment]
> > > > > > > arify when should we do
> > > > > > 
> > > > > > 
> > > > > > @bardliao As I know from HW engineer's comments, the codec driver can do io_init function only once.
> > > > > > When the bus enters clockstop1 mode, the codec just loses the context of MIPI-defined registers.
> > > > > > But, the codec could keep the settings in which set by io_init function.
> > > > > > I could double confirm whether the codec loses any settings of io_init after the enumeration.
> > > > > 
> > > > > 
> > > > > @bardliao @plbossart I had checked the context of the registers set by io_init (rt700/rt711/rt1308) after the enumeration.
> > > > > As I said, the codec will keep the context of implementation defined registers.
> > > > > I think the driver could do io_init only once when codec initialization.
> > > > 
> > > > 
> > > > Are you sure? before the enumeration starts, the master will issue a Severe Reset (Hard Reset before v1.2) with the alternating pattern of 4096+ 0/1s, and that will force the device to reset. Are you saying the imp-def registers are not affected by a Severe Reset?
> > > 
> > > 
> > > @plbossart Yes, in use case currently, the clockstopmode1 reset just affects MIPI-defined registers.
> > > The imp-def registers will keep the context except for power-on-reset or internal reset command.
> > > Do you have any playback/recording problems after the enumeration without more io_init calls?
> > 
> > 
> > I was not talking about the clockstopmode1 reset, but the "Bus Reset" (Table 39 of SoundWire spec 1.1, Table 44 of spec 1.2). On S3 resume, the Master starts with a Bus Reset and that _should_ have the effect of resetting everything. This Bus Reset helps make sure the lines are in the right electrical level and that the codec is in a known state. When we implement clock-stop, we will use it only for pm_runtime. We have no plans to support clockstopmode1 for now.
> 
> I checked with HW engineer again. The rt700/rt711/rt1308 follow SdW spec. 1.1. So I focus on SdW spec. 1.1.
> According to Table 39, BusReset belongs to the Hard of severity. It is the same with clockstopmode1 reset. The HW follows the Table 40 to make the effect of reset in slave. These effects don't include imp-def registers. Hence, I tested again with BusReset, the imp-def still keeps the context.
> If there is a sentence to indicate the imp-def needs to reset, please tell me. I can notice the HW engineer to check.

no, there is nothing in the spec that requires the imp-def registers to be reset, but on the other hand this means the Master has no ability to force the device to go back to the same state as power-on-exit. The consequence is that we should read all imp-def registers on a bus reset, or conversely re-execute the routine that sets them to hardware defaults.

**Comment by plbossart** [comment]
@shumingfan @bardliao I updated the PM support since the first pass did not fully work. The initial code did not suspend, now it does. I had to use a magic combination of setting the device active, then taking a reference and releasing it.

I also tried to move the pm_runtime init to the rt711_init() function done during the probe, as suggested by @ranj063, but that's not desirable. There are cases where we probe based on information on an ACPI device present, but the hardware never reports present, so we never suspend that non-existent device and as a result the parent does not suspend either. So I kept the use of first_init for now. Depending on the discussion on initialization and register handling, we may merge this with the hw_init flags.

**Comment by ranj063** [comment]
> I also tried to move the pm_runtime init to the rt711_init() function done during the probe, as suggested by @ranj063, but that's not desirable. There are cases where we probe based on information on an ACPI device present, but the hardware never reports present, so we never suspend that non-existent device and as a result the parent does not suspend either. So I kept the use of first_init for now. Depending on the discussion on initialization and register handling, we may merge this with the hw_init flags.

@plbossart in the cases where we probe based on ACPI device being present, wont we be calling a remove on the device if hw doesnt report present?

**Comment by plbossart** [comment]
> > I also tried to move the pm_runtime init to the rt711_init() function done during the probe, as suggested by @ranj063, but that's not desirable. There are cases where we probe based on information on an ACPI device present, but the hardware never reports present, so we never suspend that non-existent device and as a result the parent does not suspend either. So I kept the use of first_init for now. Depending on the discussion on initialization and register handling, we may merge this with the hw_init flags.
> 
> @plbossart in the cases where we probe based on ACPI device being present, wont we be calling a remove on the device if hw doesnt report present?

who's 'we'? the framework does not have a notion of timeout after which devices might be considered AWOL.
It's going to be classic to have the BIOS expose more devices than are actually present, it's convenient for BIOS folks so that a single table can represent multiple configs.

**Comment by ranj063** [comment]
> who's 'we'? the framework does not have a notion of timeout after which devices might be considered AWOL.
> It's going to be classic to have the BIOS expose more devices than are actually present, it's convenient for BIOS folks so that a single table can represent multiple configs.

@plbossart this sounds dubious. Lets say you probe the device and irrespective of whether you enable runtime PM or not, its parent device should not be able to suspend if the child device isnt runtime suspended isnt it?

**Comment by ranj063** [comment]
> @plbossart this sounds dubious. Lets say you probe the device and irrespective of whether you enable runtime PM or not, its parent device should not be able to suspend if the child device isnt runtime suspended isnt it?

@plbossart answering my own question. I was under the impression that if a child device's runtime status is "unsupported" the parent should automatically be runtime active all the time. Turns out thats not the case. 

Anyway, I think in this case, the sequence you have looks just fine

**Comment by bardliao** [comment]
> no, there is nothing in the spec that requires the imp-def registers to be reset, but on the other hand this means the Master has no ability to force the device to go back to the same state as power-on-exit. The consequence is that we should read all imp-def registers on a bus reset, or conversely re-execute the routine that sets them to hardware defaults.

I think re-execute that sets them to hardware defaults is better. We can consider it as a part of reset function.

**Comment by plbossart** [comment]
@shumingfan @bardliao I merged my code with yours. Please check if the results look ok.
I added this to integration/soundwire-latest, this branch will not compile on its own due to missing headers.

**Pull request merged**